src/coprocessor/dag/batch/executors/slow_hash_aggr_executor.rs

```diff
@@ -188,63 +181,108 @@ impl<Src: BatchExecutor> AggregationExecutorImpl<Src> for SlowHashAggregationImp
-            let group_by_result =
-                group_by_exp.eval(&mut entities.context, rows_len, src_schema, &mut input)?;
-            // Unwrap is fine because we have verified the group by expression before.
-            let group_column = group_by_result.vector_value().unwrap();
-            let field_type = group_by_result.field_type();
-            for row_index in 0..rows_len {
-                group_by_keys_offsets[row_index].push(group_by_keys[row_index].len() as u32);
-                group_column.encode(row_index, field_type, &mut group_by_keys[row_index])?;
-            }
-        }
-        // One extra offset, to be used as the end offset.
-        for row_index in 0..rows_len {
-            group_by_keys_offsets[row_index].push(group_by_keys[row_index].len() as u32);
+        let rows_len = input.rows_len();
+        let group_by_len = self.group_by_exps.len();
+        let aggr_fn_len = entities.each_aggr_fn.len();
+
+        // Decode columns with mutable input first, so subsequent access to input can be immutable
+        // (and the borrow checker will be happy)
+        ensure_columns_decoded(&context.cfg.tz, &self.group_by_exps, src_schema, &mut input)?;
+        ensure_columns_decoded(
+            &context.cfg.tz,
+            &entities.each_aggr_exprs,
+            src_schema,
+            &mut input,
+        )?;
+        assert!(self.group_by_results.is_empty());
+        assert!(self.aggr_expr_results.is_empty());
+        unsafe {
+            eval_exprs_no_lifetime(
+                context,
+                &self.group_by_exps,
+                src_schema,
+                &input,
+                &mut self.group_by_results,
+            )?;
+            eval_exprs_no_lifetime(
+                context,
+                &entities.each_aggr_exprs,
+                src_schema,
+                &input,
+                &mut self.aggr_expr_results,
+            )?;
         }
 
-        for (group_key, group_key_offsets) in group_by_keys.into_iter().zip(group_by_keys_offsets) {
-            match self.groups.entry(group_key) {
+        for row_index in 0..rows_len {
+            let offset_begin = self.group_key_buffer.len();
+            for group_by_result in &self.group_by_results {
+                match group_by_result {
+                    RpnStackNode::Vector { value, field_type } => {
+                        value.encode(row_index, field_type, &mut self.group_key_buffer)?;
+                        self.group_key_offsets.push(self.group_key_buffer.len());
+                    }
+                    // we have checked that group by cannot be a scalar
+                    _ => unreachable!(),
+                }
+            }
+            let group_key_ref_unsafe = GroupKeyRefUnsafe {
+                buffer: (&*self.group_key_buffer).into(),
+                begin: offset_begin,
+                end: self.group_key_buffer.len(),
+            };
+            let group_len = self.groups.len();
```

Comment:
No need to calculate group_len when it is not Entry::Vacant

Reply:
`HashMap::entry` takes `&mut self`. We cannot get its length when we have another mutable ref.